Add documentation (#989)
MetaMask/controllers: docs/why.md

> # Why a monorepo?

Reviewer: Thought: `why.md` isn't a very descriptive name and this focuses on a specific aspect of the architecture. Though it does seem relevant and useful as such.

Here's one thought: Merge `what.md`, `why.md`, and the *Structure* section of `how.md` into `architecture.md` , and this could be a `## Motivation` section there?

Reviewer 2: That would make sense to me. The historical context for how things used to work, and the specific problems we have, will rapidly become less relevant. Explaining the ongoing benefits is useful though.

Reviewer 2: Reminds me of the old README this package had, though that was far longer and even less relevant. There was an elaborate justification given for why we should use TypeScript over JavaScript. I'm sure people needed convincing at the time, but it felt a bit silly after we became more invested in using TS as a team.

Author: Good points! I'll try condensing the content and removing the historical angle.

Prior to November 2022, this repo was a monolithic repo, where the code for many [controllers](./what.md) lived under one roof and was published as one package (`@metamask/controllers`). This led to a few problems:

- Even if a product team wanted to use a small subset of controllers, they were forced to add the entire package to their project, which bloated the dependency tree unnecessarily.
- Occasionally, a product team tasked with carving out a new feature needed to make temporary modifications to one or more controllers during development, but this was difficult, as they were forced to fork the repo and keep it up to date.
- Beyond the controllers repo, we were maintaining code for other shared libraries in separate repos, and keeping these repos architecturally aligned with each other had grown to be painful.

> We aimed to improve these situations by splitting up `@metamask/controllers` into many packages, where each package corresponded to a single controller (except for a few cases). By doing this:

Reviewer: ```suggestion
Attempts to improve these situations were made by splitting up `@metamask/controllers` into many packages, where each package corresponded to a single controller (except for a few cases). By doing this:
```

> - Product teams could now choose which controllers they wanted to use in their project without the fear of relying on more dependencies than were necessary.

Reviewer: ```suggestion
- Users can now pick and choose use which controllers they want to use in their project without worrying about pulling in more dependencies than necessary.
```

> - Instead of forking, product teams can publish [preview builds](./preview-builds.md) for in-progress work and make use of those builds in products to test them out as they are working on new features.

Reviewer: Just for my education: Why was this facilitated by breaking apart the packages? What made that unviable for the monopackage?

Reviewer 2: I'm not sure about this sentence. Previously, controller forks weren't (to my knowledge) used for experiments that often. They were used by mobile as a medium-term way to introduce controller changes quickly, bypassing the need to make that change functional for the extension and agreeable for other teams.

Experiments were done by installing deps from git references. Which was made impractical due to the Yarn v3 `workspaces: ` protocol.

Reviewer 2: What we could list here as a benefit is that we've made it easier to fork individual controllers without everything else, reducing the maintenance burden of those forks.

Author: Yeah, this sentence doesn't make any sense as a motivation, does it 😕  Admittedly, I editorialized the monorepo kickoff document a bit. Given that we aren't able to link to that document anyway I think I will just steal what you wrote @Gudahtt. Maybe the part about forking still makes sense but I'll work it in more as a benefit than a motivation as you said.

- Although not carried out yet, in the future we hope to bring shared libraries as well as additional controllers not hosted here into this repo so that we can standardize the shape of each package much more easily than we can today.

> For a more in-depth explanation of challenges and solutions, read the [kickoff document for the monorepo project](https://docs.google.com/document/d/1G3M-lcwvfNFs3Tq4lVzhywqzYhCqPqBN7c5n8TC9sWM/edit).

Reviewer: Authwalled link - can this be published?

Reviewer 2: Hmm. I guess we could publish it. It seems a bit in-the-weeds though, not especially relevant to external contributors, especially over time.

We should at least note that it's an internal doc though (anyone with a ConsenSys account should have comment access).

Author: I think I might remove this link entirely and work in more of the text itself into this document (assuming that it still makes sense to explain the motivations behind the monorepo).
